protocols/imap/src/main/java/org/apache/james/imap/processor/base/UidMsnMapper.java

```diff
@@ -35,54 +37,101 @@
     public final static int FIRST_MSN = 1;
 
     private final HashBiMap<Integer, MessageUid> msnToUid;
+    private final ReadWriteLock readWriteLock;
 
     public UidMsnMapper() {
+        readWriteLock = new ReentrantReadWriteLock();
```

Comment:
why dit you choose a readwritelock ? What are the gains ? (complexity is higher than basic synchronize, I expect a gain, isn't it ?)